Prod deploy failed on main

## Prod deploy failed

**Repo**: alexsiri7/reli
**SHA**: `6bbe0bf8bf648900b84458357c4b589504643c5a`
**Deploy run/status**: https://github.com/alexsiri7/reli/actions/runs/25249993085
**State**: failure
**Deployed at**: 2026-05-02T10:34:40Z

Auto-filed by `pipeline-health-cron.sh`. Prod deploy is a pipeline bottleneck — archon has been fired immediately. This issue is tagged `archon:in-progress` so the pickup cron will not double-fire.

### Steps
1. Inspect deploy logs at https://github.com/alexsiri7/reli/actions/runs/25249993085
2. Identify root cause (Railway config? env var? build error? migration?)
3. Fix on a branch, land, confirm next deploy goes green

## 🔍 Investigation: Prod deploy failed on main (RAILWAY_TOKEN expired — 56th occurrence)

**Type**: `BUG` (infrastructure / secret rotation)

### Assessment

| Metric | Value | Reasoning |
|--------|-------|-----------|
| Severity | `HIGH` | Prod deploy is fully blocked at the validator step (no workaround), but no data loss / security exposure; a known recurring infra issue with a documented fix path. |
| Complexity | `LOW` | Zero code changes — a single GitHub Actions secret value (`RAILWAY_TOKEN`) must be replaced by a human admin via railway.com → repo Settings. |
| Confidence | `HIGH` | Failed run [25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) logs the exact failure (`RAILWAY_TOKEN is invalid or expired: Not Authorized`) at the `Validate Railway secrets` step; identical signature to 55 prior incidents (#742 → … → #891). |

---

### Problem Statement

The `Deploy to staging` job in run [25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) failed at the **Validate Railway secrets** step at 2026-05-02T10:34:36Z. Railway's GraphQL API responded `Not Authorized` to the `{me{id}}` validation probe — the `RAILWAY_TOKEN` GitHub Actions secret is expired or revoked. This is the **56th** RAILWAY_TOKEN expiration tracked on this repo and the **16th today**, arriving ~29 minutes after #891. The ~30-minute inter-arrival now holds across **eight** consecutive incidents (#878 → #880 → #882 → #884 → #886 → #888 → #891 → #894). Rotation tracker issue **#889** is already OPEN.

---

### Root Cause Analysis

**WHY**: Prod deploy failed on commit `6bbe0bf` at 2026-05-02T10:34:40Z.
↓ **BECAUSE**: The `Deploy to staging` workflow exited 1 at the `Validate Railway secrets` step.
&nbsp;&nbsp;Evidence: `##[error]RAILWAY_TOKEN is invalid or expired: Not Authorized`
↓ **BECAUSE**: The validator's GraphQL probe to `backboard.railway.app/graphql/v2` returned `Not Authorized`.
&nbsp;&nbsp;Evidence: `.github/workflows/staging-pipeline.yml:32-58` posts `{"query":"{me{id}}"}` and exits 1 if `.data.me.id` is missing.
↓ **ROOT CAUSE**: The `RAILWAY_TOKEN` GitHub Actions secret holds an expired/revoked Railway API token. Identical to #891, #888, #886, #884, …, #742 — all resolved by rotating the secret value via railway.com.

---

### Implementation Plan

| Step | Actor | Action |
|------|-------|--------|
| 1 | **Human admin** | Rotate `RAILWAY_TOKEN` per `docs/RAILWAY_TOKEN_ROTATION_742.md` (select **No expiration**) |
| 2 | Human admin | Re-run [run 25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) and confirm green |
| 3 | Human admin | Close #894 and #889 |

> ⚠️ **Per `CLAUDE.md` Railway Token Rotation policy, agents cannot rotate this token.** Do not create a `.github/RAILWAY_TOKEN_ROTATION_894.md` file claiming rotation is done — that is a Category 1 error.

<details>
<summary>📋 Full implementation steps</summary>

1. Log into railway.com.
2. Generate a new API token under Account Settings → Tokens. **Select "No expiration"** — short-TTL defaults are the cause of the recurring failure.
3. In GitHub: repo Settings → Secrets and variables → Actions → update `RAILWAY_TOKEN` with the new value.
4. Re-run the failed workflow at https://github.com/alexsiri7/reli/actions/runs/25249993085 (or wait for next push to `main`).
5. Confirm the next deploy goes green.
6. Close issues **#894** and **#889**.

</details>

---

### Validation

```bash
gh run rerun 25249993085
gh run watch 25249993085
```

Confirm the `Validate Railway secrets` step now reports success and subsequent deploy steps complete.

---

### Follow-up (out of scope here)

Eight consecutive ~30-minute expirations (#878 → #880 → #882 → #884 → #886 → #888 → #891 → #894) is structural. A **separate** issue / mail to mayor should evaluate: project-scoped Railway token (with `Project-Access-Token` header), workspace/account token with explicit no-expiration TTL, long-lived deploy credential, or alternate hosting. Do not bundle that work with this incident (Polecat Scope Discipline).

---

### Next Step

Human admin: follow `docs/RAILWAY_TOKEN_ROTATION_742.md`. To implement the docs-only investigation artifact: `/implement-issue 894`.

---
*Investigated by Claude • 2026-05-02T11:05:00Z*

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
@@ -20,8 +20,8 @@
 
 | File | Action | Lines |
 |------|--------|-------|
-| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md` | CREATE | +151 |
-| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md` | CREATE | +~250 |
+| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md` | CREATE | +150 |
+| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md` | CREATE | +159 |
 | `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md` | CREATE | (this file) |
 
 No source, workflow, or configuration files were modified. The artifact's

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
@@ -1,0 +1,86 @@
+# Implementation Report
+
+**Issue**: #894
+**Generated**: 2026-05-02 11:10
+**Workflow ID**: 594db19c756acf05e346a8d70e5a6a19
+
+---
+
+## Tasks Completed
+
+| # | Task | File | Status |
+|---|------|------|--------|
+| 1 | Commit investigation artifact for #894 (56th RAILWAY_TOKEN expiration, 16th today) | `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md` | ✅ |
+| 2 | Commit web-research artifact for #894 | `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md` | ✅ |
+| 3 | Write this implementation report | `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md` | ✅ |
+
+---
+
+## Files Changed
+
+| File | Action | Lines |
+|------|--------|-------|
+| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md` | CREATE | +151 |
+| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md` | CREATE | +~250 |
+| `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md` | CREATE | (this file) |
+
+No source, workflow, or configuration files were modified. The artifact's
+Implementation Plan explicitly contains zero in-repo code changes — Step 1
+("Human admin rotates the RAILWAY_TOKEN secret") is out-of-band, and Steps 2-3
+(re-run workflow, close issues) execute against GitHub, not the working tree.
+
+---
+
+## Deviations from Investigation
+
+Implementation matched the investigation exactly. No code, workflow, or config
+edits were performed because the investigation's Implementation Plan and Scope
+Boundaries both specify none.
+
+The agent did **not** create a `.github/RAILWAY_TOKEN_ROTATION_894.md` file —
+per `CLAUDE.md` → "Railway Token Rotation" and the artifact's Edge Cases
+section, that would be a Category 1 error (claiming success on an action the
+agent cannot perform).
+
+---
+
+## Validation Results
+
+| Check | Result |
+|-------|--------|
+| Type check | n/a — no source changes |
+| Tests | n/a — no source changes |
+| Lint | n/a — no source changes |
+| Artifact integrity (`investigation.md` and `web-research.md` present in `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/`) | ✅ |
+| No fabricated `RAILWAY_TOKEN_ROTATION_894.md` | ✅ |
+| Runbook still present at `docs/RAILWAY_TOKEN_ROTATION_742.md` | ✅ |
+| Validator workflow still at `.github/workflows/staging-pipeline.yml` | ✅ |
+
+The full automated check suite (`bun run type-check`, `bun test`, `bun run lint`)
+is intentionally not invoked: this PR touches only documentation under
+`artifacts/runs/`, so those checks have no signal. The artifact's Validation
+section calls for `gh run rerun 25249993085 && gh run watch 25249993085` once a
+human has rotated the token — that is post-rotation verification and runs
+outside this implementation phase.
+
+---
+
+## Hand-off
+
+The next action is **out-of-band** and belongs to a human admin with
+railway.com access:
+
+1. Mint a new Railway token (select **No expiration** if available — see
+   investigation §"Edge Cases & Risks" and web-research §"Findings 1, 4").
+2. Update the `RAILWAY_TOKEN` GitHub Actions secret.
+3. `gh run rerun 25249993085 && gh run watch 25249993085` to confirm the
+   `Validate Railway secrets` step now passes (note: `staging-pipeline.yml`
+   contains a second validator at line 149 that reuses the same secret, so a
+   single rotation greens both).
+4. Close issues **#894** and **#889** (the rotation tracker).
+
+Structural follow-up (eight consecutive ~30-minute expirations) is explicitly
+out of scope for this PR per Polecat Scope Discipline; see investigation
+§"Follow-up" for the recommendation to evaluate project-scoped tokens,
+no-expiration account/workspace tokens, or alternate hosting in a separate
+issue / mayor mail.

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md
@@ -1,0 +1,150 @@
+# Investigation: Prod deploy failed on main (RAILWAY_TOKEN expired — 56th occurrence)
+
+**Issue**: #894 (https://github.com/alexsiri7/reli/issues/894)
+**Type**: BUG (infrastructure / secret rotation)
+**Investigated**: 2026-05-02T11:05:00Z
+
+### Assessment
+
+| Metric | Value | Reasoning |
+|--------|-------|-----------|
+| Severity | HIGH | Prod deploy is fully blocked at the validator step (no workaround), but no data loss or security exposure; a known recurring infra issue with a documented fix path. |
+| Complexity | LOW | Zero code changes — a single GitHub Actions secret value (`RAILWAY_TOKEN`) must be replaced by a human admin via railway.com → repo Settings. |
+| Confidence | HIGH | Failed run [25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) logs the exact failure (`RAILWAY_TOKEN is invalid or expired: Not Authorized`) at the `Validate Railway secrets` step; identical signature to 55 prior incidents (#742 → … → #891). |
+
+---
+
+## Problem Statement
+
+The `Deploy to staging` job in run [25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) failed at the **Validate Railway secrets** step at 2026-05-02T10:34:36Z. Railway's GraphQL API responded `Not Authorized` to the `{me{id}}` validation probe — the `RAILWAY_TOKEN` GitHub Actions secret is expired or revoked. This is the **56th** RAILWAY_TOKEN expiration tracked on this repo and the **16th today**, arriving ~29 minutes after #891. The ~30-minute inter-arrival now holds across **eight** consecutive incidents (#878 → #880 → #882 → #884 → #886 → #888 → #891 → #894). Rotation tracker issue **#889** is already OPEN.
+
+---
+
+## Analysis
+
+### Root Cause / Change Rationale
+
+The `RAILWAY_TOKEN` GitHub Actions secret holds an expired/revoked Railway API token. The workflow's pre-deploy validator probes Railway's GraphQL API and exits 1 when the token is rejected. No code path or workflow change can fix this — only a human with railway.com access can mint a new token and update the GitHub secret.
+
+### Evidence Chain
+
+WHY: Prod deploy failed on commit `6bbe0bf` at 2026-05-02T10:34:40Z.
+↓ BECAUSE: The `Deploy to staging` workflow exited 1 at the `Validate Railway secrets` step.
+  Evidence: `##[error]RAILWAY_TOKEN is invalid or expired: Not Authorized` (run 25249993085 logs)
+
+↓ BECAUSE: The validator's GraphQL probe to `backboard.railway.app/graphql/v2` returned `Not Authorized`.
+  Evidence: `.github/workflows/staging-pipeline.yml:32-58` posts `{"query":"{me{id}}"}` and exits 1 if `.data.me.id` is missing.
+
+↓ ROOT CAUSE: The `RAILWAY_TOKEN` GitHub Actions secret holds an expired/revoked Railway API token. Identical to #891, #888, #886, #884, …, #742 — all resolved by rotating the secret value via railway.com.
+
+### Affected Files
+
+| File | Lines | Action | Description |
+|------|-------|--------|-------------|
+| (none) | — | — | No code changes. Fix is rotating the `RAILWAY_TOKEN` GitHub Actions secret via the runbook in `docs/RAILWAY_TOKEN_ROTATION_742.md`. |
+
+### Integration Points
+
+- `.github/workflows/staging-pipeline.yml:32` — `Validate Railway secrets` step (where the failure surfaces)
+- GitHub repo Settings → Secrets and variables → Actions → `RAILWAY_TOKEN` (where the value lives)
+- railway.com → Account Settings → Tokens (where the new token is minted)
+
+### Git History
+
+- **Recurring pattern**: #742 → … → #882 → #884 → #886 → #888 → #891 → #894
+- **Today's cadence**: 16 expirations on 2026-05-02 (#894 is the 16th); ~30-minute inter-arrival now holds for 8 consecutive incidents
+- **Implication**: Long-standing infrastructure issue (token TTL too short); behavior-of-code is correct, the secret value is the problem
+
+---
+
+## Implementation Plan
+
+### Step 1: Human admin rotates the RAILWAY_TOKEN secret
+
+**File**: (none — GitHub Actions secret, not in repo)
+**Action**: ROTATE
+
+**What to do** (per `docs/RAILWAY_TOKEN_ROTATION_742.md`):
+1. Log into railway.com.
+2. Generate a new API token under Account Settings → Tokens. **Select "No expiration"** — short-TTL defaults are the cause of the recurring failure.
+3. In GitHub: repo Settings → Secrets and variables → Actions → update `RAILWAY_TOKEN` with the new value.
+
+**Why**: The token validator in `.github/workflows/staging-pipeline.yml:32-58` probes Railway's GraphQL API and exits 1 when `Not Authorized` is returned. Only a fresh token resolves this.
+
+> ⚠️ **Per `CLAUDE.md` Railway Token Rotation policy, agents cannot rotate this token.** Do not create a `.github/RAILWAY_TOKEN_ROTATION_894.md` file claiming rotation is done — that is a Category 1 error.
+
+### Step 2: Re-run the failed workflow
+
+**Action**: VERIFY
+
+```bash
+gh run rerun 25249993085
+gh run watch 25249993085
+```
+
+**Why**: Confirm the `Validate Railway secrets` step now passes and downstream deploy steps complete.
+
+### Step 3: Close the related issues
+
+**Action**: CLEANUP
+
+Close **#894** (this issue) and **#889** (the rotation tracker) once the next deploy goes green.
+
+---
+
+## Patterns to Follow
+
+The runbook at `docs/RAILWAY_TOKEN_ROTATION_742.md` is the canonical procedure. The previous 55 investigations (most recently #891) are all variants of this same artifact — none required code changes.
+
+---
+
+## Edge Cases & Risks
+
+| Risk/Edge Case | Mitigation |
+|----------------|------------|
+| New token also short-TTL → another expiration in ~30 min | Select **No expiration** when minting (or longest available); see Follow-up. |
+| Re-run picks up cached failure | Use `gh run rerun --failed` or push a fresh commit to `main`. |
+| Agent attempts rotation and fabricates a `RAILWAY_TOKEN_ROTATION_894.md` | Explicit Category 1 prohibition in `CLAUDE.md`; this artifact reiterates it. |
+
+---
+
+## Validation
+
+### Automated Checks
+
+```bash
+gh run rerun 25249993085
+gh run watch 25249993085
+```
+
+The `Validate Railway secrets` step should report success and downstream `Deploy` steps should complete green.
+
+### Manual Verification
+
+1. Confirm the new token value is saved in GitHub Actions secrets.
+2. Watch the rerun (or next `main` push) reach the `Deploy to staging` step successfully.
+3. Confirm `https://reli-staging.up.railway.app` (or equivalent) responds.
+
+---
+
+## Scope Boundaries
+
+**IN SCOPE:**
+- Document the failure, point to the rotation runbook, hand off to the human admin.
+
+**OUT OF SCOPE (do not touch):**
+- Code changes to `.github/workflows/staging-pipeline.yml` (validator is working correctly).
+- Creating a `RAILWAY_TOKEN_ROTATION_894.md` claiming rotation is done — agents cannot rotate the token.
+- Bundling the structural fix (long-lived token, project-scoped token, alternate hosting) into this incident — that belongs in a separate issue / mail to mayor (Polecat Scope Discipline).
+
+### Follow-up (separate issue / mayor mail)
+
+Eight consecutive ~30-minute expirations (#878 → #880 → #882 → #884 → #886 → #888 → #891 → #894) is structural. Evaluate independently: project-scoped Railway token (with `Project-Access-Token` header), workspace/account token with explicit no-expiration TTL, long-lived deploy credential, or alternate hosting.
+
+---
+
+## Metadata
+
+- **Investigated by**: Claude
+- **Timestamp**: 2026-05-02T11:05:00Z
+- **Artifact**: `/home/asiri/.archon/workspaces/alexsiri7/reli/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md`

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/validation.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/validation.md
@@ -22,12 +22,13 @@ All validation checks are vacuously satisfied: the only deliverables for this ta
 
 ## Branch State
 
-**Command**: `git diff --name-only HEAD~1 HEAD`
+**Command**: `git diff --name-only main...HEAD`
 **Result**:
 
 ```
 artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
 artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md
+artifacts/runs/594db19c756acf05e346a8d70e5a6a19/validation.md
 artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md
 ```
 

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/validation.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/validation.md
@@ -1,0 +1,135 @@
+# Validation Results
+
+**Generated**: 2026-05-02 11:18
+**Workflow ID**: 594db19c756acf05e346a8d70e5a6a19
+**Status**: ALL_PASS (vacuous — investigation-only task with no code under test)
+
+---
+
+## Summary
+
+| Check | Result | Details |
+|-------|--------|---------|
+| Type check | N/A | No source-code changes in branch |
+| Lint | N/A | No source-code changes in branch |
+| Format | N/A | No source-code changes in branch |
+| Tests | N/A | No source-code changes in branch |
+| Build | N/A | No source-code changes in branch |
+
+All validation checks are vacuously satisfied: the only deliverables for this task are documentation artifacts (`investigation.md`, `web-research.md`, `implementation.md`) under `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/`. No files outside that directory were modified.
+
+---
+
+## Branch State
+
+**Command**: `git diff --name-only HEAD~1 HEAD`
+**Result**:
+
+```
+artifacts/runs/594db19c756acf05e346a8d70e5a6a19/implementation.md
+artifacts/runs/594db19c756acf05e346a8d70e5a6a19/investigation.md
+artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md
+```
+
+**Command**: `git status`
+**Result**: `nothing to commit, working tree clean`.
+
+This matches `investigation.md` § "Affected Files" (no code changes) and § "Scope Boundaries" (out of scope: workflow files, runbook, any `RAILWAY_TOKEN_ROTATION_*.md`). It also matches `implementation.md` § "Files Changed", which lists only the three artifacts above.
+
+---
+
+## Why No Source Validation Was Run
+
+Per `CLAUDE.md` § "Railway Token Rotation":
+
+> Agents cannot rotate the Railway API token. The token lives in GitHub Actions secrets (`RAILWAY_TOKEN`) and requires human access to railway.com. … Creating documentation that claims success on an action you cannot perform is a Category 1 error.
+
+Per `investigation.md` § "Scope Boundaries":
+
+> **OUT OF SCOPE (do not touch)**:
+> - Code changes to `.github/workflows/staging-pipeline.yml` (validator is working correctly).
+> - Creating a `RAILWAY_TOKEN_ROTATION_894.md` claiming rotation is done.
+> - Bundling the structural fix into this incident.
+
+Per `implementation.md` § "Validation Results":
+
+> The full automated check suite (`bun run type-check`, `bun test`, `bun run lint`) is intentionally not invoked: this PR touches only documentation under `artifacts/runs/`, so those checks have no signal.
+
+Running `lint` / `test` / `build` on an unchanged source tree would only re-validate the `origin/main` baseline that was already validated when the prior 55 investigation PRs (#742 → … → #891 → #892) merged. It would burn CI minutes for zero signal on this PR's deliverable.
+
+---
+
+## Type Check
+
+**Command**: `npm --prefix frontend run build` (includes `tsc -b`) — not run
+**Result**: N/A — no `.ts` / `.tsx` / `.py` changes in the branch.
+
+> Note: this repo has no standalone `type-check` script. TypeScript checking happens as the first step of `frontend`'s `build` script (`tsc -b && vite build`).
+
+---
+
+## Lint
+
+**Command**: `npm --prefix frontend run lint` — not run
+**Result**: N/A — no source files changed.
+
+> Backend has no lint script defined in `pyproject.toml`.
+
+---
+
+## Format
+
+**Command**: (no `format:check` script exists in `frontend/package.json`) — not run
+**Result**: N/A — no source files changed.
+
+---
+
+## Tests
+
+**Command**: `npm --prefix frontend test` (vitest) — not run
+**Result**: N/A — no source files changed; the prior `origin/main` test run on the predecessor PR (#892, SHA `b4b2daa`) is the relevant baseline.
+
+---
+
+## Build
+
+**Command**: `npm --prefix frontend run build` — not run
+**Result**: N/A — no source files changed.
+
+---
+
+## Files Modified During Validation
+
+None.
+
+---
+
+## Validation That *Did* Apply
+
+The investigation already specifies the validation that matters for this task — and it is not source-code validation, it is a post-rotation runtime check that only a human with railway.com access can perform:
+
+```bash
+# After human rotates RAILWAY_TOKEN per docs/RAILWAY_TOKEN_ROTATION_742.md:
+gh run rerun 25249993085
+gh run watch 25249993085
+# Expect: Validate Railway secrets ✅; Deploy to staging ✅
+```
+
+That step is owned by the human operator and recorded here for traceability only.
+
+Artifact-integrity checks (which *do* apply to this PR's deliverable) all pass:
+
+| Check | Result |
+|-------|--------|
+| `investigation.md` present in `artifacts/runs/594db19c756acf05e346a8d70e5a6a19/` | ✅ |
+| `web-research.md` present in same directory | ✅ |
+| `implementation.md` present in same directory | ✅ |
+| No fabricated `.github/RAILWAY_TOKEN_ROTATION_894.md` | ✅ |
+| Runbook still present at `docs/RAILWAY_TOKEN_ROTATION_742.md` | ✅ |
+| Validator workflow still at `.github/workflows/staging-pipeline.yml` (line 32) | ✅ |
+
+---
+
+## Next Step
+
+Continue to `archon-finalize-pr` to update the PR and mark it ready for human review (with the explicit hand-off that the human admin must rotate `RAILWAY_TOKEN` and rerun [run 25249993085](https://github.com/alexsiri7/reli/actions/runs/25249993085) before issues #894 and #889 can close).

--- a/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md
+++ b/artifacts/runs/594db19c756acf05e346a8d70e5a6a19/web-research.md
@@ -1,0 +1,159 @@
+# Web Research: fix #894
+
+**Researched**: 2026-05-02T (UTC)
+**Workflow ID**: 594db19c756acf05e346a8d70e5a6a19
+**Issue**: [#894 — Prod deploy failed on main](https://github.com/alexsiri7/reli/issues/894)
+
+---
+
+## Summary
+
+Issue #894 is the **56th** instance of `RAILWAY_TOKEN is invalid or expired: Not Authorized` failing the `Validate Railway secrets` step in `.github/workflows/staging-pipeline.yml`. The recurring nature suggests a structural problem, not a one-off rotation. Web research surfaces two relevant findings: (1) Railway has multiple token types (account, workspace, project, OAuth) with different scopes, and the `RAILWAY_TOKEN` env var name has a strict mapping — only **project tokens** are accepted in `RAILWAY_TOKEN`, while account/workspace tokens belong in `RAILWAY_API_TOKEN`; (2) the long-term fix for "rotate every N days" pain in CI is OIDC-based short-lived credentials, but Railway does not currently offer GitHub-OIDC federation, so the practical mitigation is choosing a non-expiring token type and adding pre-flight rotation alerts.
+
+Per repo policy (`CLAUDE.md` → Railway Token Rotation), agents cannot rotate the token. This research is meant to inform the human who performs rotation about whether the token type itself can be changed to reduce recurrence.
+
+---
+
+## Findings
+
+### 1. Railway has four distinct token types — choosing the right one matters
+
+**Source**: [Railway Public API — Tokens](https://docs.railway.com/integrations/api)
+**Authority**: Official Railway documentation
+**Relevant to**: Root cause — wrong token type may explain recurring expirations
+
+**Key Information**:
+
+- Four token types exist: **Account token**, **Workspace token**, **Project token**, **OAuth access token**.
+- **Account token**: access to all your resources and workspaces (broadest scope, tied to a single human user).
+- **Workspace token**: access to a single workspace; doc says "Best For: Team CI/CD, shared automation" and explicitly states "you can share this token with your teammates."
+- **Project token**: scoped to a specific environment within a project; only authenticates requests to that environment.
+- The currently configured secret in `staging-pipeline.yml:34,62,151,179` is named `RAILWAY_TOKEN` but the workflow queries `{me{id}}` against `backboard.railway.app/graphql/v2` — `me` is an account-level field, which means the token in use is **not** a pure project token; it must have account scope.
+
+---
+
+### 2. `RAILWAY_TOKEN` env var name only accepts project tokens with the CLI
+
+**Source**: [RAILWAY_TOKEN invalid or expired — Railway Help Station](https://station.railway.com/questions/railway-token-invalid-or-expired-59011e20)
+**Authority**: Railway community Q&A (official help station)
+**Relevant to**: Possible misconfiguration — env var/token-type mismatch
+
+**Key Information**:
+
+- Direct community quote: *"RAILWAY_TOKEN now only accepts project token, if u put the normal account token...it literally says 'invalid or expired'"*.
+- Account tokens go in `RAILWAY_API_TOKEN`, project tokens go in `RAILWAY_TOKEN` (per [Using the CLI](https://docs.railway.com/guides/cli)).
+- Having both `RAILWAY_TOKEN` and `RAILWAY_API_TOKEN` set simultaneously can cause conflicts.
+- Reli's workflow does **not** use the Railway CLI — it makes raw `curl` calls to the GraphQL API with `Authorization: Bearer $RAILWAY_TOKEN`. For raw API calls, the env var name is just shell convention; what matters is the token's scope. So this constraint does **not** apply to Reli's current setup, but it would matter if the workflow ever migrated to the CLI.
+
+---
+
+### 3. Token expiration policies — what the docs do and do not say
+
+**Source**: [Login & Tokens | Railway Docs](https://docs.railway.com/integrations/oauth/login-and-tokens), [Public API | Railway Docs](https://docs.railway.com/integrations/api)
+**Authority**: Official Railway documentation
+**Relevant to**: Whether *any* token type avoids expiration
+
+**Key Information**:
+
+- OAuth **access tokens** explicitly expire after **1 hour**; OAuth **refresh tokens** have a **1-year lifetime from issuance** and are rotated.
+- For **account, workspace, and project tokens**, the public docs **do not state an explicit expiration policy**. Community reports suggest these are intended to be long-lived but in practice users do see "expired" errors that may be tied to other invalidation events (token revocation on user permission changes, workspace membership changes, security incidents, or undocumented TTLs).
+- January 2026 Railway incident report ([blog.railway.com/p/incident-report-january-26-2026](https://blog.railway.com/p/incident-report-january-26-2026)) describes auth failures from rate-limit pressure (GitHub OAuth 2k/hr cap exceeded during sign-up surge) — *this is a different failure mode* but worth mentioning since transient `Not Authorized` could also stem from upstream incidents rather than true expiration.
+
+**Gap**: There is no authoritative source stating "workspace tokens never expire." The closest thing is the workspace-token "Best For Team CI/CD" framing, which implies long-lived intent.
+
+---
+
+### 4. The systemic answer for CI rotation pain is OIDC — but Railway doesn't support it
+
+**Source**: [GitHub OIDC docs](https://docs.github.com/en/actions/concepts/security/openid-connect), [StepSecurity best practices](https://www.stepsecurity.io/blog/github-actions-security-best-practices)
+**Authority**: Official GitHub docs + recognized security vendor
+**Relevant to**: Long-term elimination of token-rotation toil
+
+**Key Information**:
+
+- OIDC eliminates long-lived secrets entirely: the cloud provider issues a short-lived token per workflow run, automatically scoped and rotated.
+- GitHub Actions natively supports OIDC for **AWS, Azure, GCP, HashiCorp Vault**, and other providers offering official login actions.
+- Railway is **not** in the list of providers with documented GitHub-OIDC federation as of this research. So OIDC is not currently a drop-in option for Reli's deploy flow.
+- Alternative: external secret manager (HashiCorp Vault, Infisical, Doppler) with auto-rotation hooks — overkill for this single-token problem.
+
+---
+
+### 5. Workspace token is the doc-recommended type for shared/team CI
+
+**Source**: [Railway Public API — Tokens](https://docs.railway.com/integrations/api), [Token for GitHub Action — Help Station](https://station.railway.com/questions/token-for-git-hub-action-53342720)
+**Authority**: Official docs + Railway community
+**Relevant to**: Concrete change that may reduce recurrence
+
+**Key Information**:
+
+- Doc explicitly labels workspace tokens "Best For: Team CI/CD, shared automation."
+- Help-station thread: *"use a Railway API token scoped to the user account, not a project token, for the GitHub Action, set in the `RAILWAY_API_TOKEN` environment variable... when creating the token in the accounts page, you must leave the workspace blank to create an account-scoped token; setting it to your default workspace creates a workspace-scoped token instead."*
+- Trade-off: a workspace token is shared, so when the workspace owner changes membership/permissions or rotates credentials, deploys break. An account token tied to a specific human is even more fragile (breaks if that human leaves or rotates their password).
+
+---
+
+## Code Examples
+
+Reli's current validation pattern (from `staging-pipeline.yml:49-58`):
+
+```bash
+# From .github/workflows/staging-pipeline.yml
+RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
+  -H "Authorization: Bearer $RAILWAY_TOKEN" \
+  -H "Content-Type: application/json" \
+  -d '{"query":"{me{id}}"}')
+if ! echo "$RESP" | jq -e '.data.me.id' > /dev/null 2>&1; then
+  MSG=$(echo "$RESP" | jq -r '.errors[0].message // "could not reach Railway API or token rejected"')
+  echo "::error::RAILWAY_TOKEN is invalid or expired: $MSG"
+fi
+```
+
+The `{me{id}}` probe confirms the secret is being used as an **account-scoped** token (project tokens have no `me` field). This means the token type is account or workspace — not project.
+
+Reference workflow pattern using project token + Railway CLI (from [bervProject/railway-deploy](https://github.com/bervProject/railway-deploy)):
+
+```yaml
+# From https://github.com/bervProject/railway-deploy
+- run: npm i -g @railway/cli
+- run: railway up --service=backend
+  env:
+    RAILWAY_TOKEN: ${{ secrets.RAILWAY_TOKEN }}  # project token here
+```
+
+---
+
+## Gaps and Conflicts
+
+- **Gap**: Railway public docs do not state explicit expiration TTLs for account, workspace, or project tokens. Empirically, this repo has rotated 56 times — strong signal that *something* is invalidating the token periodically (TTL, security policy, or revocation on workspace events). Without official confirmation, we cannot assert any specific token type "never expires."
+- **Gap**: No public Railway docs on GitHub-OIDC federation — appears unsupported as of May 2026.
+- **Conflict**: Some sources say `RAILWAY_TOKEN` requires a project token; the Reli workflow uses an account-scoped token in `RAILWAY_TOKEN` and it has historically *worked* (just expired). Reconciliation: the env-var name only matters when using the **Railway CLI**, which enforces the mapping. Reli uses raw `curl` against the GraphQL API where the bearer token is just a bearer token — the var-name constraint does not apply.
+- **Possible non-rotation cause**: Some recurring "expired" reports correlate with Railway-side rate-limiting or auth incidents, not actual token expiry. With 56 occurrences, rate-limiting is unlikely to be the dominant cause, but it could account for a subset.
+
+---
+
+## Recommendations
+
+Based on research:
+
+1. **Verify which token type is in `RAILWAY_TOKEN` today.** The probe `{me{id}}` works → token is account or workspace scope. If it's an **account token tied to a single user**, switching to a **workspace token** scoped to the team workspace is the doc-recommended improvement and removes the dependency on one human's account state.
+2. **Don't migrate the env-var name unless adopting the Railway CLI.** The workflow uses raw GraphQL; renaming `RAILWAY_TOKEN` → `RAILWAY_API_TOKEN` would only matter if/when CLI usage is added.
+3. **Avoid claiming OIDC will fix this.** Railway has no documented GitHub-OIDC federation. Suggesting OIDC migration would be misleading.
+4. **For this specific issue (#894):** follow `CLAUDE.md` Railway Token Rotation policy — file the issue (already done, that's #894), point the human at `docs/RAILWAY_TOKEN_ROTATION_742.md`, do **not** create another `RAILWAY_TOKEN_ROTATION_*.md` "claim of completion" file. Implementation work that *would* reduce recurrence (switch token type, add pre-expiry monitoring) is out of scope for the rotation itself and should be tracked separately if the human wants it.
+5. **Pre-expiry monitoring** is feasible and in-scope for an agent: a scheduled workflow that runs `{me{id}}` daily and opens an issue when it starts failing, rather than waiting for a deploy to fail. (Repo already has `railway-token-health.yml` — suggesting this monitoring may exist; worth verifying.)
+
+---
+
+## Sources
+
+| # | Source | URL | Relevance |
+|---|--------|-----|-----------|
+| 1 | Railway Public API (Tokens) | https://docs.railway.com/integrations/api | Token types and intended use |
+| 2 | Railway CLI Login & Tokens | https://docs.railway.com/integrations/oauth/login-and-tokens | OAuth token TTLs (1h access / 1y refresh) |
+| 3 | Using the CLI | https://docs.railway.com/guides/cli | `RAILWAY_TOKEN` vs `RAILWAY_API_TOKEN` env-var mapping |
+| 4 | RAILWAY_TOKEN invalid or expired (Help Station) | https://station.railway.com/questions/railway-token-invalid-or-expired-59011e20 | Token-type/env-var mismatch as common root cause |
+| 5 | Token for GitHub Action (Help Station) | https://station.railway.com/questions/token-for-git-hub-action-53342720 | Account-scoped token guidance for shared CI |
+| 6 | Railway incident report Jan 28-29 2026 | https://blog.railway.com/p/incident-report-january-26-2026 | Upstream auth failures unrelated to token expiry |
+| 7 | GitHub Actions OIDC docs | https://docs.github.com/en/actions/concepts/security/openid-connect | Long-term zero-rotation pattern (not yet supported by Railway) |
+| 8 | StepSecurity — GitHub Actions security best practices | https://www.stepsecurity.io/blog/github-actions-security-best-practices | Industry best-practice framing for secret rotation |
+| 9 | bervProject/railway-deploy | https://github.com/bervProject/railway-deploy | Reference workflow using Railway CLI + project token |
+| 10 | Using GitHub Actions with Railway (blog) | https://blog.railway.com/p/github-actions | Official Railway-published GitHub Actions guidance |

